packages/editor/src/components/document-outline/index.js

```diff
@@ -113,105 +143,182 @@ export default function DocumentOutline( {
-			blocks: getBlocks(),
-			isTitleSupported: postType?.supports?.title ?? false,
-		};
-	} );
+	const { blocks, title, isTitleSupported, isEditingTemplate } = useSelect(
+		( select ) => {
+			const { getBlocks } = select( blockEditorStore );
+			const { getEditedPostAttribute } = select( editorStore );
+			const { getPostType } = select( coreStore );
+			const postType = getPostType( getEditedPostAttribute( 'type' ) );
 
+			return {
+				title: getEditedPostAttribute( 'title' ),
+				blocks: getBlocks(),
+				isTitleSupported: postType?.supports?.title ?? false,
+				isEditingTemplate: postType?.slug === 'wp_template',
+			};
+		}
+	);
 	const prevHeadingLevelRef = useRef( 1 );
 
-	const headings = computeOutlineHeadings( blocks );
-	if ( headings.length < 1 ) {
-		return (
-			<div className="editor-document-outline has-no-headings">
-				<EmptyOutlineIllustration />
-				<p>
-					{ __(
-						'Navigate the structure of your document and address issues like empty or incorrect heading levels.'
-					) }
-				</p>
-			</div>
-		);
-	}
+	const outlineElements = computeOutlineElements( blocks, isEditingTemplate );
+	const headings = outlineElements.filter(
+		( item ) => item.type === 'heading'
+	);
+	const mainElements = outlineElements.filter(
+		( item ) => item.type === 'main'
+	);
 
 	// Not great but it's the simplest way to locate the title right now.
 	const titleNode = document.querySelector( '.editor-post-title__input' );
 	const hasTitle = isTitleSupported && title && titleNode;
-	const countByLevel = headings.reduce(
-		( acc, heading ) => ( {
-			...acc,
-			[ heading.level ]: ( acc[ heading.level ] || 0 ) + 1,
-		} ),
-		{}
-	);
+
+	// Count the number of headings and nested headings by level, to determine if there are multiple H1s.
+	const countByLevel = outlineElements.reduce( ( acc, element ) => {
+		if ( element.type === 'heading' ) {
+			acc[ element.level ] = ( acc[ element.level ] || 0 ) + 1;
+		}
+		if (
+			element.type === 'main' &&
+			element.children &&
+			element.children.length > 0
+		) {
+			element.children.forEach( ( child ) => {
+				if ( child.type === 'heading' ) {
+					acc[ child.level ] = ( acc[ child.level ] || 0 ) + 1;
+				}
+			} );
+		}
+
+		return acc;
+	}, {} );
 	const hasMultipleH1 = countByLevel[ 1 ] > 1;
 
-	return (
-		<div className="document-outline">
-			<ul>
-				{ hasTitle && (
+	const documentOutlineItems = ( { item } ) => {
+		if ( isEditingTemplate && item.type === 'main' ) {
+			return (
+				<li key={ item.clientId }>
 					<DocumentOutlineItem
-						level={ __( 'Title' ) }
+						level={ __( 'Main' ) }
 						isValid
-						onSelect={ onSelect }
-						href={ `#${ titleNode.id }` }
 						isDisabled={ hasOutlineItemsDisabled }
+						href={ `#block-${ item.clientId }` }
+						onSelect={ () => {
+							selectBlock( item.clientId );
+							onSelect?.();
+						} }
 					>
-						{ title }
+						{ incorrectMainTag }
 					</DocumentOutlineItem>
```

Comment:
If I don't close this off before listing the children, the SelectBlock does not work on the children.
The result is that the children are not indented.